File: frontend/src/types/cards.ts
// Mirrors the Pydantic models in backend/models.py

export type VerdictStatus = "keep" | "situational" | "reconsider";
export type CreditTier = "easy" | "plan" | "niche";
export type InsuranceLevel = "strong" | "good" | "mid" | "none";
export type TimelineEventType = "add" | "cut" | "neutral" | "future";

export interface Verdict {
  status: VerdictStatus;
  text: string;
  short_tag?: string | null;
}

/** A promotional intro-APR period, on purchases or balance transfers. `rate`
 * is almost always "0%" but kept as a string, not a boolean, since a small
 * number of real-world promos use a non-zero teaser rate. */
export interface IntroApr {
  rate: string;
  months: number;
}

/** The current sign-up offer, quoted verbatim from the issuer's own current
 * terms — offers change frequently and are structured differently across
 * issuers (points vs. cash, single vs. tiered spend thresholds), so `bonus`/
 * `requirement` stay free text like `effective_cost`/`variable_apr`
 * elsewhere. `estimated_value` is this card's own best redemption cpp
 * applied to the bonus, when it's a points/miles amount — null for cash
 * offers, where the bonus amount already is the value. */
export interface WelcomeBonus {
  bonus: string;
  requirement: string;
  estimated_value: string | null;
}

interface CardBase {
  id: string;
  name: string;
  issuer: string;
  network: string;
  points_program: string;
  accent_color: string;
  annual_fee: number;
  effective_cost: string;
  verdict: Verdict;
  /** Set on an unsecured card whose secured counterpart has identical earn
   * rates — the secured card is hidden from catalog listings in favor of
   * this one, and this id is used to link to it as "also available". */
  secured_variant_id: string | null;
  /** The inverse, set on the secured card itself (computed server-side,
   * never stored in that card's own data) — lets its own detail page link
   * back to the unsecured primary it's hidden in favor of, since it's no
   * longer reachable by browsing catalog listings. */
  is_secured_variant_of: string | null;
  /** Shared, non-unique id for a real-world transferable points pool (e.g.
   * Chase Freedom Flex/Unlimited pool into a Sapphire Preferred/Reserve
   * account and inherit its redemption value). Every card sharing this id
   * is a peer, not a primary/duplicate pair like secured_variant_id — used
   * by Top Pick's "My Cards" ranking to value a held flat-rate card at the
   * best cpp among every other held card in the same pool. */
  points_pool_id: string | null;
  /** True only for the card whose own account a pooled balance actually
   * redeems through (e.g. Sapphire Preferred/Reserve, Strata Premier/
   * Elite) — false for a feeder card that only reaches full value once
   * pooled with one of these. A feeder must never be valued off another
   * feeder sharing its pool id with no receiver present — real-world
   * transfer-partner access requires an active premium account in the
   * mix, not just any two linked cards. */
  points_pool_receiver: boolean;
  /** Whether this card's official_url carries affiliate tracking
   * (commission-earning). False for every card today — the catalog has no
   * affiliate program yet. Drives CardDetailPage's AffiliateDisclosure,
   * and — just as importantly — is carried on CardSummary specifically so
   * Top Pick's ranking (computeTopPicks, which only ever sees CardSummary)
   * can be proven to ignore monetization status entirely: see the
   * ranking-integrity regression test in topPickCategories.test.ts. */
  is_affiliate_link: boolean;
  /** Sourced from the issuer's own current terms, not guessed — null means
   * this card hasn't been audited yet (distinct from confirmed-absent),
   * during the incremental per-issuer rollout of this data. Drives the
   * Compare tab's "0% Intro APR" / "Balance Transfer" Category chips: a
   * promotional balance-transfer offer IS an intro-APR period applied to
   * balance transfers, so there's no separate "has balance transfer" flag
   * that could drift out of sync with this. */
  intro_apr_purchases: IntroApr | null;
  intro_apr_balance_transfers: IntroApr | null;
  /** True = card charges a foreign transaction fee. Null = not yet audited.
   * Drives the "No Foreign Transaction Fee" Category chip (only matches
   * when this is explicitly `false`, not merely falsy/null). */
  foreign_transaction_fee: boolean | null;
  /** Derived server-side from status_perks (any perk mentioning "lounge") —
   * unlike the three fields above, this doesn't need per-card research, so
   * there's no "not yet audited" state, just true/false. */
  has_lounge_access: boolean;
}

export interface EarnCategorySummary {
  category: string;
  multiplier: string;
  /** Whether this is the card's flat "everything else" rate — the only
   * reliable Catch-All signal, since that rate is phrased dozens of
   * different ways in category free text. */
  is_base: boolean;
}

/** Returned by GET /api/cards */
export interface CardSummary extends CardBase {
  total_easy_credits: number;
  total_max_credits: number;
  /** Category + multiplier pairs only (not full EarnRate objects) — enough to
   * derive spend-category filter chips (Dining, Gas, ...) AND rank matches by
   * multiplier catalog-wide, without fetching every card's full detail. */
  categories: EarnCategorySummary[];
  /** Best cents-per-point across this card's redemption options (1.0 for a
   * flat cash-back card). Lets the Top Pick page rank by effective earn
   * rate (multiplier × best_cpp) instead of comparing raw multipliers
   * across incompatible currencies. */
  best_cpp: number;
}

export interface EarnRate {
  emoji: string;
  multiplier: string;
  category: string;
  highlight: boolean;
  is_base: boolean;
}

export interface RedemptionOption {
  method: string;
  cpp: number;
  best: boolean;
}

export interface Points {
  currency: string;
  redemption_options: RedemptionOption[];
  per_100k: string;
  note: string;
}

export type TransferPartnerType = "airline" | "hotel";

export interface TransferPartner {
  name: string;
  type: TransferPartnerType;
  ratio: string;
  notes?: string | null;
}

export interface TransferPartners {
  airline_count: number;
  hotel_count: number;
  highlight: string;
  recent_changes: string;
  partners?: TransferPartner[];
}

export interface Credit {
  id: string;
  name: string;
  subtitle: string;
  max_annual: number;
  default_value: number;
  tier: CreditTier;
  removed: boolean;
  description: string;
  tips: string[];
}

export interface Insurance {
  coverage: string;
  detail: string;
  level: InsuranceLevel;
}

export interface StatusPerk {
  name: string;
  strength: number;
  note: string;
}

export interface Service {
  name: string;
  detail: string;
}

export interface AdditionalCardBenefit {
  text: string;
  included: boolean;
}

export interface AdditionalCardOption {
  name: string;
  fee: string;
  is_free: boolean;
  benefits: AdditionalCardBenefit[];
}

export interface AdditionalCards {
  title: string;
  options: AdditionalCardOption[];
  note: string;
}

export interface TimelineEvent {
  date: string;
  type: TimelineEventType;
  badge: string;
  text: string;
}

/** Returned by GET /api/cards/:id */
export interface Card extends CardBase {
  official_url?: string | null;
  earn_rates: EarnRate[];
  earn_note: string;
  points: Points;
  transfer_partners: TransferPartners;
  credits: Credit[];
  insurance: Insurance[];
  protection_note: string;
  rental_note: string;
  status_perks: StatusPerk[];
  services: Service[];
  additional_cards: AdditionalCards;
  timeline: TimelineEvent[];
  /** Ongoing (post-intro, or from day one if there's no intro offer) rate
   * ranges and fees, quoted verbatim from the issuer's own Pricing & Terms
   * table rather than decomposed into numbers — real ranges are tied to
   * creditworthiness ("19.24%-29.99% Variable") and nothing here computes
   * with them, only displays them. Detail-only (not on CardSummary) —
   * purely informational, not a Compare-tab filter driver like
   * intro_apr_purchases/intro_apr_balance_transfers. Null = not yet
   * audited. */
  variable_apr: string | null;
  balance_transfer_apr: string | null;
  balance_transfer_fee: string | null;
  /** The actual rate when foreign_transaction_fee is true (e.g. "3%") — the
   * boolean still drives the "No Foreign Transaction Fee" chip unchanged,
   * this is just the detail-page-level specific number. */
  foreign_transaction_fee_rate: string | null;
  /** Round-3 rates/fees (#12) — same verbatim-string, detail-only,
   * null-means-not-yet-audited treatment as variable_apr above.
   * pay_over_time_fee only applies to issuers offering a "pay a purchase
   * off over time for a fixed fee instead of interest" feature (Amex's Pay
   * Over Time, Chase Pay Over Time) — null for cards with no such feature
   * at all, not just unaudited ones. */
  cash_advance_apr: string | null;
  penalty_apr: string | null;
  /** When the penalty APR kicks in (e.g. "after a payment more than 60 days
   * late") — kept separate from the rate since issuers phrase the trigger
   * very differently and it isn't always present even when the rate is. */
  penalty_apr_trigger: string | null;
  pay_over_time_fee: string | null;
  late_payment_fee: string | null;
  returned_payment_fee: string | null;
  returned_check_fee: string | null;
  /** The current sign-up offer — see WelcomeBonus. Detail-only, same
   * not-yet-audited treatment as the round-3 fields above. */
  welcome_bonus: WelcomeBonus | null;
}

File: backend/services/cards.py
from sqlalchemy.orm import Session, joinedload, selectinload

from backend.db import session_scope
from backend.db_models import AdditionalCardOption as DbAdditionalCardOption
from backend.db_models import CardModel, CardTransferPartner, CreditModel, InsuranceBenefit
from backend.models import (
    AdditionalCardBenefit,
    AdditionalCardOption,
    AdditionalCards,
    Card,
    CardSummary,
    Credit,
    EarnCategorySummary,
    EarnRate,
    Insurance,
    Points,
    RedemptionOption,
    Service,
    StatusPerk,
    TimelineEvent,
    TransferPartner,
    TransferPartners,
    Verdict,
)

# Full detail load — every relationship the Card response needs, in as few
# round trips as possible. selectinload for one-to-many (separate IN query per
# collection, no row-multiplication), joinedload for the many-to-one lookups
# (single extra join, always present). Chained loader options take the target
# class's attribute, not a string, so a rename shows up as a type error here
# instead of a silent runtime miss.
_DETAIL_OPTIONS = (
    joinedload(CardModel.issuer),
    joinedload(CardModel.network),
    joinedload(CardModel.points_program),
    selectinload(CardModel.earn_rates),
    selectinload(CardModel.redemption_options),
    selectinload(CardModel.credits).selectinload(CreditModel.tier),
    selectinload(CardModel.credits).selectinload(CreditModel.tips),
    selectinload(CardModel.insurance_benefits).joinedload(InsuranceBenefit.coverage_type),
    selectinload(CardModel.status_perks),
    selectinload(CardModel.services),
    selectinload(CardModel.additional_card_options).selectinload(DbAdditionalCardOption.benefits),
    selectinload(CardModel.timeline_events),
    selectinload(CardModel.transfer_partners).joinedload(CardTransferPartner.loyalty_program),
)

# Summary load — only what CardSummary actually needs: the scalar card fields,
# the three lookup names, credits (to total up easy/max credit values),
# earn_rates (category labels only, for spend-category filter chips), and
# redemption_options (cents-per-point only, for best_cpp). Deliberately skips
# insurance/timeline/etc. so GET /api/cards doesn't drag in relationships it
# never reads.
_SUMMARY_OPTIONS = (
    joinedload(CardModel.issuer),
    joinedload(CardModel.network),
    joinedload(CardModel.points_program),
    selectinload(CardModel.credits).selectinload(CreditModel.tier),
    selectinload(CardModel.earn_rates),
    selectinload(CardModel.redemption_options),
)


def _to_card(c: CardModel, is_secured_variant_of: str | None = None) -> Card:
    return Card(
        id=c.slug,
        name=c.name,
        issuer=c.issuer.name,
        network=c.network.name,
        points_program=c.points_program.name,
        accent_color=c.accent_color,
        annual_fee=c.annual_fee_cents // 100,
        effective_cost=c.effective_cost_label,
        official_url=c.official_url,
        verdict=Verdict(
            status=c.verdict_status, text=c.verdict_text, short_tag=c.verdict_short_tag
        ),
        earn_rates=[
            EarnRate(
                emoji=r.emoji or "",
                multiplier=r.multiplier_label,
                category=r.category,
                highlight=r.is_highlight,
                is_base=r.is_base,
            )
            for r in c.earn_rates
        ],
        earn_note=c.earn_note or "",
        points=Points(
            currency=c.points_program.name,
            redemption_options=[
                RedemptionOption(method=o.method, cpp=o.cents_per_point, best=o.is_best)
                for o in c.redemption_options
            ],
            per_100k=c.points_per_100k_label,
            note=c.points_note or "",
        ),
        transfer_partners=TransferPartners(
            airline_count=c.transfer_airline_count,
            hotel_count=c.transfer_hotel_count,
            highlight=c.transfer_highlight or "",
            recent_changes=c.transfer_recent_changes or "",
            partners=[
                TransferPartner(
                    name=p.loyalty_program.name,
                    type=p.partner_type,
                    ratio=p.transfer_ratio,
                    notes=p.notes,
                )
                for p in c.transfer_partners
            ],
        ),
        credits=[
            Credit(
                id=cr.slug,
                name=cr.name,
                subtitle=cr.subtitle,
                max_annual=cr.max_annual_cents // 100,
                default_value=cr.default_value_cents // 100,
                tier=cr.tier.code,
                removed=cr.is_removed,
                description=cr.description,
                tips=[t.tip_text if not t.is_warning else f"warn::{t.tip_text}" for t in cr.tips],
            )
            for cr in c.credits
        ],
        insurance=[
            Insurance(coverage=i.coverage_type.name, detail=i.detail, level=i.level)
            for i in c.insurance_benefits
        ],
        protection_note=c.protection_note or "",
        rental_note=c.rental_note or "",
        status_perks=[
            StatusPerk(name=p.name, strength=p.strength, note=p.note) for p in c.status_perks
        ],
        services=[Service(name=s.name, detail=s.detail) for s in c.services],
        additional_cards=AdditionalCards(
            title=c.additional_cards_title or "",
            options=[
                AdditionalCardOption(
                    name=o.name,
                    fee=o.fee_label,
                    is_free=o.is_free,
                    benefits=[
                        AdditionalCardBenefit(text=b.benefit_text, included=b.is_included)
                        for b in o.benefits
                    ],
                )
                for o in c.additional_card_options
            ],
            note=c.additional_cards_note or "",
        ),
        timeline=[
            TimelineEvent(date=e.date_label, type=e.event_type, badge=e.badge, text=e.description)
            for e in c.timeline_events
        ],
        secured_variant_id=c.secured_variant_id,
        is_secured_variant_of=is_secured_variant_of,
        points_pool_id=c.points_pool_id,
        points_pool_receiver=c.points_pool_receiver,
    )


def _to_card_summary(c: CardModel, is_secured_variant_of: str | None = None) -> CardSummary:
    active_credits = [cr for cr in c.credits if not cr.is_removed]
    easy_total = sum(
        cr.default_value_cents // 100 for cr in active_credits if cr.tier.code == "easy"
    )
    max_total = sum(cr.max_annual_cents // 100 for cr in active_credits)
    return CardSummary(
        id=c.slug,
        name=c.name,
        issuer=c.issuer.name,
        network=c.network.name,
        points_program=c.points_program.name,
        accent_color=c.accent_color,
        annual_fee=c.annual_fee_cents // 100,
        effective_cost=c.effective_cost_label,
        verdict=Verdict(
            status=c.verdict_status, text=c.verdict_text, short_tag=c.verdict_short_tag
        ),
        total_easy_credits=easy_total,
        total_max_credits=max_total,
        categories=[
            EarnCategorySummary(
                category=r.category, multiplier=r.multiplier_label, is_base=r.is_base
            )
            for r in c.earn_rates
        ],
        best_cpp=max((o.cents_per_point for o in c.redemption_options), default=0.0),
        secured_variant_id=c.secured_variant_id,
        is_secured_variant_of=is_secured_variant_of,
        points_pool_id=c.points_pool_id,
        points_pool_receiver=c.points_pool_receiver,
    )


def _query(session: Session, *options):
    return (
        session.query(CardModel)
        .filter(CardModel.is_active.is_(True))
        .options(*options)
        .order_by(CardModel.card_id)
    )


def get_card(card_id: str) -> Card | None:
    with session_scope() as session:
        card = (
            session.query(CardModel)
            .filter(CardModel.slug == card_id, CardModel.is_active.is_(True))
            .options(*_DETAIL_OPTIONS)
            .one_or_none()
        )
        if not card:
            return None
        # Reverse lookup: is this card itself hidden in favor of it, i.e. is
        # some other card's secured_variant_id pointing at this one? Only
        # ever matches for a secured card, never stored on its own row.
        primary = (
            session.query(CardModel.slug)
            .filter(CardModel.secured_variant_id == card_id, CardModel.is_active.is_(True))
            .scalar()
        )
        return _to_card(card, is_secured_variant_of=primary)


def get_card_summaries() -> list[CardSummary]:
    with session_scope() as session:
        cards = _query(session, *_SUMMARY_OPTIONS).all()
        # Same reverse lookup as get_card, but computed once for the whole
        # catalog instead of a per-card query, since this path already loads
        # every card in one pass.
        reverse = {c.secured_variant_id: c.slug for c in cards if c.secured_variant_id}
        return [_to_card_summary(c, is_secured_variant_of=reverse.get(c.slug)) for c in cards]
